File: tests/bloom-filter.js
'use strict';

const protocol = process.env.MNEMONIST_TRANSPORT === 'protocol' ? require('../adapter/_protocol.js') : null;

const LN2_SQUARED = Math.LN2 * Math.LN2;

function mul32(a, b) {
  return (a & 0xffff) * b + ((((a >>> 16) * b) & 0xffff) << 16) & 0xffffffff;
}

function sum32(a, b) {
  return (a & 0xffff) + (b >>> 16) + ((((a >>> 16) + b) & 0xffff) << 16) & 0xffffffff;
}

function murmurhash3(seed, data) {
  let hash = seed;
  let i;
  for (i = 0; i <= data.length - 4; i += 4) {
    let k1 = data[i] | (data[i + 1] << 8) | (data[i + 2] << 16) | (data[i + 3] << 24);
    k1 = mul32(k1, 0xcc9e2d51);
    k1 = (k1 << 15) | (k1 >>> 17);
    k1 = mul32(k1, 0x1b873593);
    hash ^= k1;
    hash = (hash << 13) | (hash >>> 19);
    hash = mul32(hash, 5);
    hash = sum32(hash, 0x6b64e654);
  }

  let k1 = 0;
  if ((data.length & 3) === 3) k1 ^= data[i + 2] << 16;
  if ((data.length & 3) >= 2) k1 ^= data[i + 1] << 8;
  if ((data.length & 3) >= 1) {
    k1 ^= data[i];
    k1 = mul32(k1, 0xcc9e2d51);
    k1 = (k1 << 15) | (k1 >>> 17);
    k1 = mul32(k1, 0x1b873593);
    hash ^= k1;
  }
  hash ^= data.length;
  hash ^= hash >>> 16;
  hash = mul32(hash, 0x85ebca6b);
  hash ^= hash >>> 13;
  hash = mul32(hash, 0xc2b2ae35);
  hash ^= hash >>> 16;
  return hash >>> 0;
}

function bytes(string) {
  const array = new Uint16Array(string.length);
  for (let i = 0; i < string.length; i++) array[i] = string.charCodeAt(i);
  return array;
}

function BloomFilter(capacityOrOptions) {
  if (!capacityOrOptions) {
    throw new Error('mnemonist/BloomFilter.constructor: a BloomFilter must be created with a capacity.');
  }
  const options = typeof capacityOrOptions === 'object' ? capacityOrOptions : { capacity: capacityOrOptions };
  if (typeof options.capacity !== 'number' || options.capacity <= 0) {
    throw new Error('mnemonist/BloomFilter.constructor: `capacity` option should be a positive integer.');
  }
  this.capacity = options.capacity;
  this.errorRate = options.errorRate || 0.005;
  if (typeof this.errorRate !== 'number' || options.errorRate <= 0) {
    throw new Error('mnemonist/BloomFilter.constructor: `errorRate` option should be a positive float.');
  }
  this._inner = protocol ? protocol.create('bloom-filter', [this.capacity, this.errorRate]) : null;
  this.clear();
}

BloomFilter.prototype.clear = function() {
  if (protocol) {
    this._inner.call('clear', []);
    this.hashFunctions = this._inner.call('hashFunctions', []);
    this.data = new Uint8Array(this._inner.call('data', []));
    return;
  }
  const bits = -this.capacity * Math.log(this.errorRate) / LN2_SQUARED;
  const length = bits / 8 | 0;
  this.hashFunctions = length * 8 / this.capacity * Math.LN2 | 0;
  this.data = new Uint8Array(length);
};

BloomFilter.prototype.add = function(string) {
  if (protocol) {
    this._inner.call('add', [string]);
    this.data = new Uint8Array(this._inner.call('data', []));
    return this;
  }
  const array = bytes(string);
  for (let seed = 0; seed < this.hashFunctions; seed++) {
    const index = murmurhash3((seed * 0xfba4c795) & 0xffffffff, array) % (this.data.length * 8);
    this.data[index >> 3] |= 1 << (7 & index);
  }
  return this;
};

BloomFilter.prototype.test = function(string) {
  if (protocol) return this._inner.call('test', [string]);
  const array = bytes(string);
  for (let seed = 0; seed < this.hashFunctions; seed++) {
    const index = murmurhash3((seed * 0xfba4c795) & 0xffffffff, array) % (this.data.length * 8);
    if (!(this.data[index >> 3] & (1 << (7 & index)))) return false;
  }
  return true;
};

BloomFilter.prototype.toJSON = function() {
  return this.data;
};

BloomFilter.from = function(iterable, options) {
  if (!options) {
    options = iterable.length || iterable.size;
    if (typeof options !== 'number') {
      throw new Error("BloomFilter.from: could not infer the filter's capacity. Try passing it as second argument.");
    }
  }
  const filter = new BloomFilter(options);
  for (const value of iterable) filter.add(value);
  return filter;
};

module.exports = BloomFilter;

File: src/bloom_filter.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

const LN2_SQUARED: f64 = std::f64::consts::LN_2 * std::f64::consts::LN_2;

#[derive(Debug, Clone)]
pub struct JsBloomFilter {
    capacity: usize,
    hashes: usize,
    data: Vec<u8>,
}

impl JsBloomFilter {
    pub fn new(capacity: usize, error_rate: f64) -> Result<Self, String> {
        if capacity == 0 {
            return Err("mnemonist/BloomFilter.constructor: `capacity` option should be a positive integer.".into());
        }
        if !error_rate.is_finite() || error_rate <= 0.0 {
            return Err("mnemonist/BloomFilter.constructor: `errorRate` option should be a positive float.".into());
        }
        let length = ((-(capacity as f64) * error_rate.ln() / LN2_SQUARED) / 8.0) as usize;
        let hashes = ((length * 8) as f64 / capacity as f64 * std::f64::consts::LN_2) as usize;
        Ok(Self { capacity, hashes, data: vec![0; length] })
    }

    pub fn capacity(&self) -> usize { self.capacity }
    pub fn hashes(&self) -> usize { self.hashes }
    pub fn data(&self) -> &[u8] { &self.data }
    pub fn clear(&mut self) { self.data.fill(0); }

    pub fn add(&mut self, value: &str) {
        for seed in 0..self.hashes {
            let index = murmurhash3((seed as u32).wrapping_mul(0xfba4c795), value) as usize % (self.data.len() * 8);
            self.data[index >> 3] |= 1 << (7 & index);
        }
    }

    pub fn test(&self, value: &str) -> bool {
        (0..self.hashes).all(|seed| {
            let index = murmurhash3((seed as u32).wrapping_mul(0xfba4c795), value) as usize % (self.data.len() * 8);
            self.data[index >> 3] & (1 << (7 & index)) != 0
        })
    }
}

fn murmurhash3(seed: u32, value: &str) -> u32 {
    let data: Vec<u16> = value.encode_utf16().collect();
    let mut hash = seed;
    let mut i = 0;
    while i + 3 < data.len() {
        let mut k1 = data[i] as u32 | (data[i + 1] as u32) << 8 | (data[i + 2] as u32) << 16 | (data[i + 3] as u32) << 24;
        k1 = k1.wrapping_mul(0xcc9e2d51).rotate_left(15).wrapping_mul(0x1b873593);
        hash ^= k1;
        hash = sum32(hash.rotate_left(13).wrapping_mul(5), 0x6b64e654);
        i += 4;
    }
    let mut k1 = 0;
    match data.len() & 3 {
        3 => k1 ^= (data[i + 2] as u32) << 16,
        _ => {}
    }
    if data.len() & 3 >= 2 { k1 ^= (data[i + 1] as u32) << 8; }
    if data.len() & 3 >= 1 {
        k1 ^= data[i] as u32;
        k1 = k1.wrapping_mul(0xcc9e2d51).rotate_left(15).wrapping_mul(0x1b873593);
        hash ^= k1;
    }
    hash ^= data.len() as u32;
    hash ^= hash >> 16;
    hash = hash.wrapping_mul(0x85ebca6b);
    hash ^= hash >> 13;
    hash = hash.wrapping_mul(0xc2b2ae35);
    hash ^ (hash >> 16)
}

fn sum32(a: u32, b: u32) -> u32 {
    (a & 0xffff)
        .wrapping_add(b >> 16)
        .wrapping_add((((a >> 16).wrapping_add(b)) & 0xffff) << 16)
}

#[derive(Debug, Clone)]
pub struct BloomFilter {
    bits: Vec<bool>,
    hashes: usize,
    inserted: usize,
}

impl BloomFilter {
    pub fn new(capacity: usize, error_rate: f64) -> Self {
        let capacity = capacity.max(1);
        let error_rate = error_rate.clamp(0.0001, 0.9999);
        let m = (-(capacity as f64) * error_rate.ln() / 2f64.ln().powi(2)).ceil() as usize;
        let k = ((m as f64 / capacity as f64) * 2f64.ln()).ceil() as usize;

        Self {
            bits: vec![false; m.max(1)],
            hashes: k.max(1),
            inserted: 0,
        }
    }

    pub fn capacity(&self) -> usize {
        self.bits.len()
    }

    pub fn hashes(&self) -> usize {
        self.hashes
    }

    pub fn size(&self) -> usize {
        self.inserted
    }

    pub fn clear(&mut self) {
        self.bits.fill(false);
        self.inserted = 0;
    }

    pub fn add<T: Hash>(&mut self, value: &T) {
        let indexes = self.indexes(value);
        for index in indexes {
            self.bits[index] = true;
        }
        self.inserted += 1;
    }

    pub fn contains<T: Hash>(&self, value: &T) -> bool {
        self.indexes(value)
            .into_iter()
            .all(|index| self.bits[index])
    }

    fn indexes<T: Hash>(&self, value: &T) -> Vec<usize> {
        let h1 = hash_with_seed(value, 0);
        let h2 = hash_with_seed(value, 1).max(1);
        (0..self.hashes)
            .map(|i| (h1.wrapping_add((i as u64).wrapping_mul(h2)) as usize) % self.bits.len())
            .collect()
    }
}

fn hash_with_seed<T: Hash>(value: &T, seed: u64) -> u64 {
    let mut hasher = DefaultHasher::new();
    seed.hash(&mut hasher);
    value.hash(&mut hasher);
    hasher.finish()
}
